significant_strikes.py

```diff
@@ -0,0 +1,500 @@
+from dateutil.relativedelta import relativedelta
+import swifter
+import pandas as pd
+
+class SignificantStrikes():
+    """
+    This class provides functionalities to calculate various significant strike statistics for UFC fighters.
+    """
+
+    def __init__(self) -> None: 
+        """
+        Initializes the SignificantStrikes class
+        """
+
+        self.default_val = 0
+
+    def create_significant_strikes_feats(self, df):
+        """
+        Creates a dataframe with added features for significant strikes and their differentials.
+
+        Args:
+            df (pd.DataFrame): The original dataframe containing fight data.
+
+        Returns:
+            pd.DataFrame: A dataframe with additional columns for significant strike features and differentials.
+        """
+        # Generate column names for significant strikes and differentials
+        col_names = self.create_col_names_significant_strikes()
+        col_names_differential = self.create_col_names_differential_significant_strikes()
+
+        # Copy the input dataframe and calculate significant strike features
+        input_df = df.copy()
+        result_features = pd.DataFrame(columns=col_names)
+        result_features[col_names] = input_df.swifter.progress_bar(True).apply(lambda row: self.calculate_significant_strikes(input_df, row['fighter_a_id'], row['fighter_b_id'], row.name, col_names), axis=1)
+
+        # Combine the input dataframe with the calculated features
+        result_df = pd.concat([input_df, result_features], axis=1)
```

Comment:
This might be a better way to do it rather than directly inserting the new columns into the original frame via the function - I don't remember but I think I did this for some functions in my Tennis model does this get rid of the warnings when running it? If it does we might have to change all the other issues to also create features this way.

Reply:
Yeah it gets rid of the PerformanceWarning and the SettingWithCopyWarning.

Reply:
🔥